### PR TITLE
Updates to token security policy issue #715

### PR DESCRIPTION
**Pull request by fabdrol**
This PR fixes (is supposed to fix) the security handling of server according to spec. @sbender9 could you do a review? I've tested the REST part of the policy against https://github.com/fabdrol/signalk-js-client/tree/sdk - which works now.

### CODE REVIEW

**Comment by fabdrol** [comment]
@sbender9 I need to resolve the issue raised in #717 as well, before reviewing/merging

**Comment by tkurki** [comment]
The diff is a mess, can’t tell what actual changes you have done.

Please make the cleanup changes in separate commits.

**Comment by tkurki** [comment]
=== can now be enforced with eslint - care to do that globally?

**Comment by tkurki** [comment]
... as well as no-unused.

File: lib/tokensecurity.js [hunk 1]
```diff
@@ -618,7 +617,8 @@ module.exports = function (app, config) {
     }
 
     if (!token) {
-      token = req.cookies && req.cookies.JAUTHENTICATION
+      token =
+        req.cookies && (req.cookies.SK_COOKIE || req.cookies.JAUTHENTICATION)
```
Comment (tkurki) [inline lib/tokensecurity.js]: Why the additional cookie name? How about making it configurable?

Comment (sbender9) [reply lib/tokensecurity.js]: SK_COOKIE was in the current PR that's open in specification. It is no longer there, so please revert this change. (or make it configurable)

**Comment by fabdrol** [comment]
@tkurki I was thinking about that, but may warrant some more discussion. I would propose moving to Standard JS, integrating linters and cleaning up the entire project. That should improve readability of the code, which should make contributing easier. It will also make contributing more time-consuming, and the job itself is quite time-consuming. Separate issue?

**Comment by tkurki** [comment]
Well, you started with more changes than strictly necessary ;-) 

One of these, please:
- separate commits for style changes, renames and the actual change
- only the actual change in this PR
- separate PRs for global formatting and renames, only actual change in this PR
not
- a single PR with a single commit that has formatting changes, renames and the actual change

FYI there is auto formatting with a commit hook with `prettier-standard` and you can `npm run format` and `npm run lint`.

File: lib/tokensecurity.js (hunk 1 above)
Comment (fabdrol) [reply lib/tokensecurity.js]: Okay, it's gone

**Comment by fabdrol** [comment]
@tkurki I reverted the style/strict equality/unused code changes for now.

**Comment by sbender9** [comment]
I don't there's actually any thing left here. `type` has been removed from the current PR.

The only change from there needed now is that `JWT` in the headers is now `Bearer`, if you make that change please keep backwards compatibility.

I would suggest to wait until that PR is approved though...

**Comment by fabdrol** [comment]
@sbender9 is the `statusCode` (#717) still an issue that I should incorporate? I saw you're doing something on that too.

**Comment by sbender9** [comment]
Yeah, I did PRs for those already.

**Comment by fabdrol** [comment]
Okay, in that case this can be merged if @tkurki is happy.

**Comment by fabdrol** [comment]
@sbender9 okay, but then we need changes to the actual policy: without type client doesn't know how to send request AND server won't authenticate if the kind isn't JWT (this line, I believe: https://github.com/SignalK/signalk-server-node/blob/d2fe5e26dc06c47d764d8ee4ca1fb466431440be/lib/tokensecurity.js#L617)

**Comment by sbender9** [comment]
The latest spec says that we always use "BEARER", "JWT" goes away. (https://github.com/SignalK/specification/blob/fix-security/gitbook-docs/security.md)

So no need for type.

Out code needs updated to support BEARER (and JWT for backwards compatibility)

**Comment by fabdrol** [comment]
Okay, I'll make those changes

**Comment by fabdrol** [comment]
@sbender9 changes have been made. @tkurki can you complete your review? I believe it's outdated ;-). Then we can merge.

File: lib/tokensecurity.js [hunk 2]
```diff
@@ -787,8 +798,16 @@ module.exports = function (app, config) {
...
-          token = header.substring('JWT '.length)
+        if (typeof header === 'string' && header.trim() !== '') {
+          token = header
+        }
+
+        if (typeof token === 'string' && token.includes('Bearer')) {
+          token = token.replace('Bearer', '').trim()
+        }
+
+        if (typeof token === 'string' && token.includes('JWT')) {
```
Comment (tkurki) [inline lib/tokensecurity.js]: If the token value just happens to include the string `JWT` or `Bearer` the auth will fail.

File: settings/volare-file-settings.json [hunk 3]
```diff
@@ -5,15 +5,15 @@
...
   "pipedProviders": [
     {
       "id": "nmeaFromFile",
       "pipeElements": [
         {
           "type": "providers/filestream",
           "options": {
-            "filename": "samples/plaka.log"
+            "filename": "samples/plaka.log",
+            "providerId": "nmeaFromFile"
```
Comment (tkurki) [inline settings/volare-file-settings.json:16]: Why the unrelated changes?

File: lib/tokensecurity.js (hunk 2 above)
Comment (sbender9) [reply lib/tokensecurity.js]: Yes, need to make sure it “starts with” them.

Comment (sbender9) [reply lib/tokensecurity.js]: Also this would leave a space before the token, I assume validation would fail?

Comment (fabdrol) [reply lib/tokensecurity.js]: @sbender9 no, that's what the `trim()` is for. @tkurki good point, didn't think that could be a case!

File: settings/volare-file-settings.json (hunk 3 above)
Comment (fabdrol) [reply settings/volare-file-settings.json:16]: Not sure, actually. I didn't touch those files. Could have been a result of d2fe5e2

File: lib/tokensecurity.js (hunk 2 above)
Comment (fabdrol) [reply lib/tokensecurity.js]: @tkurki changed, please review.

File: settings/volare-file-settings.json (hunk 3 above)
Comment (tkurki) [reply settings/volare-file-settings.json:16]: I don't think so. If you had this in the shared org repo I could fix small stuff like this myself instead of writing about it...

Comment (fabdrol) [reply settings/volare-file-settings.json:16]: Not really sure what the issue is here, btw. File seems unchanged to me

File: lib/tokensecurity.js (hunk 2 above)
Comment (tkurki) [reply lib/tokensecurity.js]: This may be nitpicking, but aren't we interested in removing `JWT ` and `Bearer ` WITH the trailing space from header values that have them? If somebody sends just a plain token value with no type separated by a space this code will now remove some data from the token value if it just happens to start with `Bearer` or `JWT`?

Comment (fabdrol) [reply lib/tokensecurity.js]: @tkurki I am not sure any token would have a trailing or starting space as part of the actual token, seems very error prone to me. But indeed, current code does not support that case. Reason I went for this solution is that, if somebody has some weird error in their client code that causes the token to be sent _without_ space after `Bearer` or with multiple spaces it'll still work. 

A different way to do it would be to simply see if whatever the client sent matches the token the server knows (using `.includes(knownToken)`). That way we don't care about the prefix and any formatting errors that may be there. 

I suppose the questions should be:

* Do we want to be character-strict about the token? In this case we shouldn't remove anything, and simply check if `sendToken === Bearer knownToken`. 
* Do we actually care about the type of token? If no, we could just do the `includes()` test

Current solution sits between these two. I personally think it is fine, since it does what we want it to do: check that there is a token and that the token is fine. And we know that the node server never sends a token with beginning/trailing spaces.

Comment (tkurki) [reply lib/tokensecurity.js]: As you can see from the code it does not verify the incoming token by comparing strings. We must be able to extract the actual token value to decode it.

I would just remove the strings `'Bearer '` and `'JWT '` from the beginning of the incoming value if they are there, just like [the current code does with JWT](https://github.com/SignalK/signalk-server-node/blob/master/lib/tokensecurity.js#L614-L615). I don't like adding extra leniency for missing or multiple spaces.

**Comment by tkurki** [comment]
Closing for inactivity, feel free to reopen.
